leader/clique/async_korach_moran_zaks/async_korach_moran_zaks.go

```diff
@@ -0,0 +1,371 @@
+				}
+			} else {
+				received_message.Index = received_index
+				state.Queue = append(state.Queue, received_message)
+			}
+		}
+	} else if state.Phase == ask_message.Phase && state.King == ask_message.King {
+		if ask_index != state.Overlord {
+			//we were the vassal of a king who became vassal of another King
+			//that King send us the ASK message
+			//the UPDATE message came before the ASK message
+			send(ask_index, createMessage(YOUR_CITIZEN, state.Phase, state.King))
+		}
+	}
+}
+
+// Some citizen accepted to become our vassal - response to our ASK from good old times when we were still a king
+func processAcceptOld(node lib.Node, received_index int, received_message message) {
+	state := getState(node)
+	state.Vassals[received_index] = received_index
+	node.SendMessage(received_index, createMessage(UPDATE, state.Phase, state.King))
+	setState(node, &state)
+}
+
+// Our old King is now a citizen so our overlord is now our vassal
+func processAcceptNew(node lib.Node, received_index int, received_message message) {
+	state := getState(node)
+
+	state.Vassals[received_index] = received_index
+	//the new overlord might have been our vassal before
+	_, ok := state.Vassals[state.AskedBefore]
+	if ok {
+		delete(state.Vassals, state.AskedBefore)
+	}
+	state.Overlord = state.AskedBefore
+
+	defer setState(node, &state)
+
+	node.SendMessage(state.Overlord, createMessage(ACCEPT, state.Phase, state.King))
+
+	for {
+		received_index, received_message = receiveMessage(node)
+		message_type := received_message.Flag
+		if message_type == UPDATE {
+			setState(node, &state)
+			processUpdate(node, received_message)
+			state = getState(node)
+			return
+		} else {
+			received_message.Index = received_index
+			state.Queue = append(state.Queue, received_message)
+		}
+	}
+}
+
+// UPDATE message is only sent when something changes
+func processUpdate(node lib.Node, received_message message) {
+	state := getState(node)
+	state.Phase = received_message.Phase
+	state.King = received_message.King
+
+	for v := range state.Vassals {
+		node.SendMessage(v, createMessage(UPDATE, state.Phase, state.King))
+	}
+	setState(node, &state)
+}
+
+/************************************************************/
+/*                         	 MAIN                           */
+/************************************************************/
+
+func initialize(node lib.Node) {
+	n := node.GetSize() - 1
+	new_state := state{}
+	new_state.Rivals = make(map[int]int)
+	new_state.Vassals = make(map[int]int)
+	for i := 0; i < n; i++ {
+		new_state.Rivals[i] = i
+	}
+	new_state.Phase = 0
+	new_state.Active = true
+	new_state.Overlord = node.GetIndex()
+	new_state.King = node.GetIndex()
+	new_state.Queue = make([]message, 0)
+	setState(node, &new_state)
+}
+
+func run(node lib.Node) {
+	node.StartProcessing()
+	initialize(node)
+	new_state := getState(node)
+	new_state.Phase = 0
+	setState(node, &new_state)
+	mainKing(node)
```

Comment:
Please refactor it in the way that there is more-or-less one for loop with every iteration: receive message, read state, process, send x messages to other nodes, finish iteration

In general, a structure like run() in leader/undirected_ring/async_itai_rodeh/async_itai_rodeh.go would be appreciated.

Reply:
Done